content/blog/next-level-iac-breakpoint-debugging/index.md

```diff
@@ -0,0 +1,317 @@
+``` 
+
+### Step 4: Setting a debugging breakpoint 
+
+Now we can try [setting a breakpoint][vs-code-breakpoints] in our code. Navigate to a line within the code and hit the `F9` key. You'll see a small red dot appear next to the line indicating the breakpoint. You can then run the program again using `F5`. When the execution of the code reaches the line it will pause there.
+
+At this point you can navigate around the code [inspecting the state of the local variables][vs-code-data-inspection], like you would in any other debugging session. There's also the "Debug" console which provides a [REPL][vs-code-debug-repl] environment to navigate those variables and anything else that is in scope. You can run arbitrary functions here and see their output. 
+
+When you're ready for the program to continue on, hit `F5` once again and the program will resume operation, either until the program exits or until it reaches the next breakpoint. Optionally you can step through line-by-line using the `F10` key. VS Code provides a variety of [debug actions][vs-code-debug-actions] once the sessions is established.
+
+## Considerations and Limitations
+
+Breakpoint debugging is a powerful tool for inspecting the state of your Pulumi program. However, there are some important limitations to consider:
+
+* **Pulumi programs are only part of the story:** This technique will let you inspect the program you authored, but it doesn't give access to what's happening with the Pulumi backend, the providers, or anything else outside of the program itself. You might find yourself needing to dig deeper to resolve a problem with the stack as a whole. Luckily there are other techniques we can use for that, which we will delve into in an upcoming post.
+
+* **Inputs and Outputs are asynchronous futures:** If you've used Pulumi before you will have encountered [`Input<T>`][input-docs] and [`Output<T>`][output-docs] values, which are a core function of the tool. However, at the time that the debugger hits a breakpoint on those values, they are likely not populated with their final value. That means you won't be able to inspect them in the debugger very well. Instead, you will need to add an [`apply(...)`][apply-docs] or [`all(..)`][all-docs] block to the code, and place the breakpoint on the materialized value within that block. This follows all the same rules as accessing those variables in code in a Pulumi program. 
+
+* **Limited integration with VS Code:** Unfortunately our technique of running Pulumi in wrapper script as a backgrounded pre-launch task in VS Code leaves some things to be desired. It is different enough from the use case that VS Code had in mind for these features that making things "just work" is a little difficult. For example, we can't easily report on errors in the "Problems" tab. Sometimes VS Code can't tell that the program finished and so you'll have to manually stop the debugging session after Pulumi has completed.
```

Comment:
can you rephrase this to be more clear. i don't really understand what is being said here.

Reply:
> Sometimes VS Code can't tell that the program finished and so you'll have to manually stop the debugging session after Pulumi has completed.

This happens 100% of the time for me, yeah. I figured this was the expected behavior, but I guess not?

It does sound like this is worth explaining in more detail, though. Like how is it "different from the use case VS Code had in mind"? Does it have something to do with the fact that the Node process is spawned by Pulumi?

Reply:
It's more about the difference between a "launch" and "attach" foreground task. Launch will run a program, attach will wait to attach to a program... So to launch the pulumi CLI we run it as a preLaunchTask (because you can't launch and then attach in the same launch config, so it needs to run it "preLaunch"), but if you don't say it's a background task, it will then have two wait conditions... waiting to attach and waiting for prelaunchtask to complete. that means you have to run as a background task (e.g. indicating that it runs forever like a service and not wait on it). Then, once it's running, the launch config will wait to attach to a debugger. but since that's a background task that can start/stop multiple times, instead of attaching then exiting once the thing it attaches to is complete, it attaches, and then goes into a wait/retry loop to try to reconnect with another background process. if it hits that timeout it will exit on its own. otherwise you have to manually exit the debug mode. Other programs behave in a way that is consistent with this model (like `tsc-watch`), which is what VS Code's implementation is designed for.

Frankly, this is all too complicated for this post which is already fairly long. I wanted to call it out without diving into the details and distracting from the "hero narrative" of Pulumi being debuggable by talking about all the ways the VS Code's debugging implementation makes it not fit perfectly into that model.

Maybe I should just remove this bullet point entirely?